### PR TITLE
User/stwum/integrate fandango json parser into tango simlib

### PR DESCRIPTION
**Pull request by samtwum**

### CODE REVIEW

File: tango_simlib/tango_sim_generator.py [hunk 1]
```diff
@@ -504,6 +508,8 @@ def get_device_class(sim_data_files):
             parser_instance = get_parser_instance(data_file)
         elif extension in [".json"] and len(sim_data_files) < 2:
             parser_instance = get_parser_instance(data_file)
+        elif extension in [".fgo"] and len(sim_data_files) <2:
```
Comment (kmadisa) [inline tango_simlib/tango_sim_generator.py]: You do not need another `elif..` block here. You can have the first `elif` block like this `..in [".fgo", ".json"] and ...`.

Comment (kmadisa) [inline tango_simlib/tango_sim_generator.py]: You do not need another `elif..` block here. You can have the first `elif` block like this `..in [".fgo", ".json"] and ...`.

Comment (samtwum) [reply tango_simlib/tango_sim_generator.py]: Oh yeah. Thanks

File: tango_simlib/tango_sim_generator.py [hunk 2]
```diff
@@ -392,6 +393,9 @@ def get_parser_instance(sim_datafile):
     elif extension in [".xml"]:
         parser_instance = SDDParser()
         parser_instance.parse(sim_datafile)
+    elif extension in [".fgo"]:
+        parser_instance = FandangoExportDeviceParser()
+        parser_instance.parse(sim_datafile)
     return parser_instance
 
 def configure_device_model(sim_data_file=None, test_device_name=None):
```
Comment (kmadisa) [inline tango_simlib/tango_sim_generator.py]: Update the docstring to be more generic instead referring to just the """...xmi file.."""

**Comment by kmadisa** [comment]
@ajoubertza I noticed that the two classes are testing the same thing for both devices and had suggested to @samtwum to rather create a base class with all the common functionality and have the test subclasses to override the `setUp` and the `setUpClassWithCleanup` methods for the different config files used.

File: tango_simlib/tests/test_fandango_json_parser.py [hunk 3]
```diff
@@ -89,12 +89,12 @@ class GenericSetup(unittest.TestCase):
...
         super(GenericSetup, self).setUp()
         self.fandango_json_file = pkg_resources.resource_filename(
-            'tango_simlib.tests.config_files', 'database2.json')
+            'tango_simlib.tests.config_files', 'database2.fgo')
         self.fandango_parser = fandango_json_parser.FandangoExportDeviceParser()
         self.fandango_parser.parse(self.fandango_json_file)
 
 
-class TestFandangoJsonParser(GenericSetup):
+class test_FandangoJsonParser(GenericSetup):
```
Comment (SKAJohanVenter) [inline tango_simlib/tests/test_fandango_json_parser.py]: For interest sake, why the change to `test_` ?

Comment (samtwum) [reply tango_simlib/tests/test_fandango_json_parser.py]: Oh it's on page 15 of the CAM - Python style guide for naming a test class.

File: tango_simlib/tests/test_dish.py [hunk 4]
```diff
@@ -1,3 +1,9 @@
+######################################################################################### 
+# Author: cam@ska.ac.za                                                                 #
+# Copyright 2018 SKA South Africa (http://ska.ac.za/)                                   #
+#                                                                                       #
+# BSD license - see LICENSE.txt for details                                             #
```
Comment (ajoubertza) [inline tango_simlib/tests/test_dish.py]: Thanks for adding all these copyright blocks.

File: tango_simlib/tests/test_tango_sim_generator.py [hunk 5]
```diff
@@ -154,3 +162,122 @@ def test_sim_control_device_attribute_change(self):
         self.sim_control_device.pause_active = True
         setattr(self.sim_control_device, 'last_val', input_value)
         self.assertEqual(self.sim_device.temperature, input_value)
+
+class test_TangoSimGenDeviceIntegrationForFandangoFile(ClassCleanupUnittestMixin,
+    unittest.TestCase):
```
Comment (ajoubertza) [inline tango_simlib/tests/test_tango_sim_generator.py]: - [ ] Don't think the indentation here isn't quite pep8 compliant.  Here are some options:
```
class test_TangoSimGenDeviceIntegrationForFandangoFile(ClassCleanupUnittestMixin,
                                                       unittest.TestCase)
```
or
```
class test_TangoSimGenDeviceIntegrationForFandangoFile(
    ClassCleanupUnittestMixin, unittest.TestCase)
```

File: tango_simlib/tests/test_tango_sim_generator.py [hunk 6]
```diff
@@ -154,3 +162,122 @@ def test_sim_control_device_attribute_change(self):
...
+        db_info_list = db_info.split('\n')
+        num_properties = 0
+        for line in db_info_list:
+            if 'Device properties defined' in line:
+                num_properties = line.split('=')[-1]
+        return int(num_properties)
+
+    def test_initial_device_properties(self):
+        """Test initial device properties added to the tangoDB"""
+        expected_count = 1
```
Comment (ajoubertza) [inline tango_simlib/tests/test_tango_sim_generator.py]: The code doesn't make it clear which property we are expecting.  The test this was copied from had a comment `# model_key property already present in db`.  

- [ ] Suggest that is used here as well (if that is indeed the same property we are getting, if not, provide a relevant comment).

File: tango_simlib/tests/config_files/database2.fgo [hunk 7]
```diff
@@ -1 +1 @@
...0.39,  0.00,  0.00,  0.00,  0.00,  0.00,  0.00,  0.00,  0.62,  0.24,  0.16,  0.00,  1.00,  0.19,  3.37,  0.00,  0.00,  0.00,  0.91,  0.00", "label": "Timing_minimum", "min_alarm": "Not specified", "events": {"per_event": {"extensions": "[]", "period": "1000"}, "ch_event": {"rel_change": "Not specified", "extensions": "[]", "abs_change": "Not specified"}, "arch_event": {"archive_period": "Not specified", "extensions": "[]", "archive_rel_change": "Not specified", "archive_abs_change": "Not specified"}}, "description": "", "data_type": "DevDouble", "format": "%6.2f", "max_alarm": "Not specified", "device": "sys/database/2", "name": "Timing_minimum", "database": "monctl:10000", "data_format": "SPECTRUM", "value": [0.29, 0.32, 0.0, 0.0, 9.23, 0.0, 0.0, 0.0, 0.39, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.62, 0.24, 0.16, 0.0, 1.0, 0.19, 3.37, 0.0, 0.0, 0.0, 0.91, 0.0], "polling": 0, "time": 1521724934.740163, "alarms": {"delta_t": "Not specified", "max_alarm": "Not specified", "min_warning": "Not specified", "extensions": "[]", "delta_val": "Not specified", "min_alarm": "Not specified", "max_warning": "Not specified"}, "model": "monctl:10000/sys/database/2/Timing_minimum"}}, "properties": {}}
```
Comment (ajoubertza) [inline tango_simlib/tests/config_files/database2.fgo]: Why were the SPECTRUM (1D) attributes removed?  I would expect them to work.  Do you know if the DSH LMC has any spectrum attributes?

File: tango_simlib/tests/test_tango_sim_generator.py [hunk 8]
```diff
@@ -154,3 +162,122 @@ def test_sim_control_device_attribute_change(self):
...
+        # be delayed by atleast 1000 ms of device server start up.
+        time.sleep(1)
+        cls.sim_device = tango.DeviceProxy(
+                '%s:%s/test/nodb/tangodeviceserver#dbase=no' % (
+                    cls.host, cls.port))
+        cls.sim_control_device = tango.DeviceProxy(
+                '%s:%s/test/nodb/tangodeviceservercontrol#dbase=no' % (
+                    cls.host, cls.port))
+        cls.addCleanupClass(cls.sub_proc.kill)
+        cls.addCleanupClass(shutil.rmtree, cls.temp_dir)
```
Comment (ajoubertza) [inline tango_simlib/tests/test_tango_sim_generator.py]: Agree with @kmadisa that there is a lot of duplication now.  Would be nice to share some of the code with tests for all parsers.  Also notice that the simdd format doesn't get tested here.  Suggest we leave the rework into a base class for another day, and another PR.

Comment (samtwum) [reply tango_simlib/tests/test_tango_sim_generator.py]: ok

File: tango_simlib/tests/config_files/database2.fgo (hunk 7 above)
Comment (samtwum) [reply tango_simlib/tests/config_files/database2.fgo]: It gives a key error when the launcher is run because it's missing max_dim_x

Comment (ajoubertza) [reply tango_simlib/tests/config_files/database2.fgo]: As discussed, we need to do some work in the fandago parser to infer values for the missing keys, since the fandango export file doesn't include them.  That can be done in a subsequent PR.
